# 01: Opening and understanding

In [12]:
library(IRdisplay)
display_html('<iframe width="560" height="315" src="https://www.youtube.com/embed/Xer1XBm3sns" frameborder="0" allowfullscreen></iframe>')

In this notebook we will go through how to open a NetCDF file in R. We will also try to understand the file and extract the data and metadata from it.

Firstly, let’s import the modules that we will work with.

In [2]:
if (!requireNamespace("RNetCDF", quietly = TRUE)) {
  install.packages("RNetCDF")
}
library(RNetCDF)

You can find the documentation for the RNetCDF package here:
https://cran.r-project.org/web/packages/RNetCDF/RNetCDF.pdf

The syntax for functions in RNetCDF is very repetitive as you will learn throughout this course, which makes it easy to learn. However, there are other libraries you can use like ncdf4.
https://cran.r-project.org/web/packages/ncdf4/ncdf4.pdf

In this tutorial, we will stick to RNetCDF.

## Importing some data from OPeNDAP

Have you heard of OPeNDAP? Try running the cell below.

OPeNDAP, which stands for “Open-source Project for a Network Data Access Protocol,” makes it easier to access and share scientific data over the internet. One advantage of using OPeNDAP is that you don’t need to download the data to use them!

In [3]:
netcdf_file <- 'https://www.ncei.noaa.gov/thredds/dodsC/noaa-global-temp-v5/NOAAGlobalTemp_v5.0.0_gridded_s188001_e202212_c20230108T133308.nc'
data <- open.nc(netcdf_file)

If we want some quick information about the file we can make an inquiry

In [4]:
file.inq.nc(data)

$ndims
[1] 4

$nvars
[1] 5

$ngatts
[1] 66

$unlimdimid
[1] NA

$format
[1] "classic"

$libvers
[1] "4.8.1 of Sep 29 2021 09:36:14 $"

The inquiry is a list of key value pairs, so we can access the values like this

In [5]:
file.inq.nc(data)['format']

$format
[1] "classic"

Though often we will want more information, so we can do this instead:

In [6]:
print.nc(data)

netcdf classic {
dimensions:
	lat = 36 ;
	lon = 72 ;
	time = 1716 ;
	z = 1 ;
variables:
	NC_FLOAT time(time) ;
		NC_CHAR time:long_name = "reference time of global temperature anomalies" ;
		NC_CHAR time:standard_name = "time" ;
		NC_CHAR time:coverage_content_type = "coordinate" ;
		NC_CHAR time:units = "days since 1800-01-01 00:00:00" ;
		NC_CHAR time:calendar = "gregorian" ;
		NC_CHAR time:axis = "T" ;
	NC_FLOAT lat(lat) ;
		NC_CHAR lat:long_name = "Latitude" ;
		NC_CHAR lat:standard_name = "latitude" ;
		NC_CHAR lat:coverage_content_type = "coordinate" ;
		NC_CHAR lat:units = "degrees_north" ;
		NC_CHAR lat:grids = "Uniform grid from -87.5 to 87.5 by 5" ;
		NC_FLOAT lat:valid_min = -87.5 ;
		NC_FLOAT lat:valid_max = 87.5 ;
		NC_CHAR lat:axis = "Y" ;
		NC_CHAR lat:_CoordinateAxisType = "Lat" ;
		NC_CHAR lat:coordinate_defines = "center" ;
	NC_FLOAT lon(lon) ;
		NC_CHAR lon:long_name = "Longitude" ;
		NC_CHAR lon:standard_name = "longitude" ;
		NC_CHAR lon:coverage_content_type = "co

Let’s break down what we see above.

A classic NetCDF file like this one can be broken down into 3 components - dimensions, variables and attributes.

* Dimensions: Define the size of the variables.
* Variables: Where the data are stored. The variables can be broken down into coordinate variables and data variables. In RNetCDF, the coordinate variables and data variables are displayed together, but if you open a NetCDF file using different software they might be separated.
    * Coordinate variables: Coordinate variables are things like longitude, latitude, time and elevation/depth. The first variable listed is *time* which has 1 dimension with the same name in brackets. So we know that the *time* variable is a 1D array with 1716 data points. It is common practice for coordinate variables to have the same name as their respective dimension, but this might not always be the case.
    * Data variables: This file has only one data variable, *anom*. This has 4 dimensions, so we can see that the data are stored in a 4D array with size 72 x 36 x 1 x 1716. 
* Attributes: The metadata describing the data. A CF-NetCDF file should have both global and variable attributes.
    * Variable attributes: Describe each variable
    * Global attributes: Describe the whole dataset (e.g. title, creator, keywords, a bounding box for the coordinates)

We will now have a closer look at how to access each of these components, starting at the bottom with attributes.

## Attributes

Accessing an attribute using RNetCDF can be done like

`att.get.nc(nc_object, variable_name, attribute_name)`

### Global attributes

For a global attribute, we assign a special variable name, *NC_GLOBAL*.

`att.get.nc(nc_object, "NC_GLOBAL", attribute_name)`

If we want to just access the *creator_name* attribute for example:

In [7]:
att.get.nc(data, "NC_GLOBAL", "creator_name")

[1] "Boyin Huang, Xungang Yin, Ph.D"

If you want more information than just the value, you can make an inquiry. See how the syntax is almost the same for 'getting' and attribute as it is for 'inquiring' about one? 

In [8]:
att.inq.nc(data, "NC_GLOBAL", "creator_name")

$id
[1] 16

$name
[1] "creator_name"

$type
[1] "NC_CHAR"

$length
[1] 30

*Conventions* is probably the most important global attribute because it tells you (and a machine) how to interpret the rest of the file.

In [9]:
att.get.nc(data, "NC_GLOBAL", "Conventions")

[1] "CF-1.6, ACDD-1.3"

*CF-1.6* refers to version 1.6 of the CF conventions, which you can find here:

https://cfconventions.org/ https://cfconventions.org/Data/cf-conventions/cf-conventions-1.6/cf-conventions.html

The CF conventions are a set of standards that define how a NetCDF file should be structured. The document linked above is extensive, but the aim is to provide a standardised way to organise many different types of data. You don’t need to read it all, but it should be your go-to place if you want to know how to structure or understand a CF-NetCDF file, or encode certain types of data.

However, the CF conventions are light on discovery metadata. Discovery metadata are metadata that can be used to find data. For example, when and where the data were collected and by whom, some keywords etc. So we also use the *ACDD* convention - The Attribute Convention for Data Discovery.

https://wiki.esipfed.org/Attribute_Convention_for_Data_Discovery_1-3

In most cases, if you want to find out what a global attribute means, you can visit the ACDD convention page above to find a description of the attribute. There are other conventions that someone might have included that you can also find online, but we recommend that you always follow the CF and ACDD conventions as a minimum when creating a NetCDF file.

The person who created this file should have read the relevant sections of these documents to make sure that the files comply with these conventions. There are also validators you can run your files by to make sure that you file is compliant with the conventions before you publish it. For example:

https://compliance.ioos.us/index.html

By following conventions, the data creator and user, human or machine, should be able to understand the data in the same way. A NetCDF file itself is not neccessarily FAIR because you could include any attributes or structure your data however you like. A CF-NetCDF file is FAIR.

### Variable attributes

Let's extract a variable attribute now from the *anom* variable:

In [10]:
att.get.nc(data, "anom", "standard_name")

[1] "surface_temperature_anomaly"

The variable name *anom* is not standardised. Fortunately, the *standard_name* variable attribute is standardised. You can find sea_water_temperature in the list of CF standard names here. 
https://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html

Hopefully, the data creator has selected the *standard_name* from the CF standard name table after reading the description for that term. You, the data user, can now also find that term and its description. This way, the data provider and the data user can share the same understanding about what the these data are. Any other CF-NetCDF file that contains the same type of data should also include the same *standard_name*. Controlled vocabularies like this one are understandable to machines too! 

The data in the file doesn’t need to be stored with the same units, but should be stored with units that are physically equivalent. For example, in this case the units degrees_C are physically equivalent to K.

## Dimensions

To extract information about a dimension, you can do this:

In [11]:
dim.inq.nc(data,'lat')

$id
[1] 0

$name
[1] "lat"

$length
[1] 36

$unlim
[1] FALSE

If you didn't know the name of the dimension, you could use the index of the variable to call it instead. Remember that 0 is always the first.

In [12]:
dim.inq.nc(data,0)

$id
[1] 0

$name
[1] "lat"

$length
[1] 36

$unlim
[1] FALSE

The *lat* dimension has 36 points. Dimensions tell you about the shape and size of your variables. In this case, we know that any variable with a dimension of only *lat* will have 36 data points - though some could be NaN. The dimension doesn't tell you anything about what the values are - that information is in the variables.

## Variables

Finally, the most important part - getting at the data!

If you know the variable name, you can get the data like this:

In [13]:
var.get.nc(data,"lat")

[1] -87.5 -82.5 -77.5 -72.5 -67.5 -62.5 -57.5 -52.5 -47.5 -42.5 -37.5 -32.5
[13] -27.5 -22.5 -17.5 -12.5  -7.5  -2.5   2.5   7.5  12.5  17.5  22.5  27.5
[25]  32.5  37.5  42.5  47.5  52.5  57.5  62.5  67.5  72.5  77.5  82.5  87.5

Or inquire about it

In [14]:
var.inq.nc(data, "lat")

$id
[1] 1

$name
[1] "lat"

$type
[1] "NC_FLOAT"

$ndims
[1] 1

$dimids
[1] 0

$natts
[1] 10

We can get just a subset of the data. For example, to get 3 values starting from the 5th value in the array, where *start* is the first value to extract and *count* is how many values:

In [15]:
var.get.nc(data, "lat", start=c(5), count=c(3))

[1] -67.5 -62.5 -57.5

There is a lot of data in the anom variable, so we won't display all of that below! But let's try and extract data for a single date. 

Let's again first select based on the index using *start* and *count. The order in the lists corresponds to the order of the dimensions that the variable has. In this case, we are extracting data for all latitudes and longitudes (so *NA* is used) and from the 1st (and only) vertical coordinate. I have selected data for an arbitrary time slice, 700 in this case.

In [20]:
lat <- var.get.nc(data, 'lat')
lon <- var.get.nc(data, 'lon')
anom <- var.get.nc(data, 'anom', start=c(NA, NA, 1, 700), count=c(NA,NA,1,1))
anom

[1] "Size of dimensions of 2D anom array"


[,1] [,2] [,3] [,4] [,5] [,6]        [,7]       [,8]        [,9]        
 [1,] NA   NA   NA   NA   NA   NA          -0.6202122 -0.28983244  0.127612844
 [2,] NA   NA   NA   NA   NA   NA          -0.4854550 -0.29407427  0.026559843
 [3,] NA   NA   NA   NA   NA   NA          -0.7393103 -0.28969392  0.041384175
 [4,] NA   NA   NA   NA   NA   NA          -0.8334374 -0.22463292  0.190988734
 [5,] NA   NA   NA   NA   NA   NA          -0.8196298 -0.20011082  0.178636834
 [6,] NA   NA   NA   NA   NA   NA          -0.5987538 -0.09562428  0.127850562
 [7,] NA   NA   NA   NA   NA   NA          -0.4492744 -0.14773342  0.020136885
 [8,] NA   NA   NA   NA   NA   NA                  NA          NA  0.009670788
 [9,] NA   NA   NA   NA   NA   NA                  NA          NA -0.158918560
[10,] NA   NA   NA   NA   NA   NA                  NA -0.64733446 -0.345341444
[11,] NA   NA   NA   NA   NA   NA                  NA -0.67590743 -0.373588830
[12,] NA   NA   NA   NA   NA   NA                  NA -0.52490014 -0.261390358
[13,] NA   NA   NA   NA   NA   NA                  NA -0.44366845 -0.299334317
[14,] NA   NA   NA   NA   NA   NA                  NA -0.44151002 -0.365199476
[15,] NA   NA   NA   NA   NA   NA                  NA -0.37778386 -0.301796347
[16,] NA   NA   NA   NA   NA   NA                  NA -0.36512563 -0.313503146
[17,] NA   NA   NA   NA   NA   NA                  NA          NA -0.343594402
[18,] NA   NA   NA   NA   NA   NA                  NA          NA -0.523310781
[19,] NA   NA   NA   NA   NA   NA                  NA          NA -0.708950460
[20,] NA   NA   NA   NA   NA   NA                  NA -0.83585399 -0.982789934
[21,] NA   NA   NA   NA   NA   NA                  NA -1.08239102 -1.135215282
[22,] NA   NA   NA   NA   NA   NA                  NA -1.20429385 -1.073588252
[23,] NA   NA   NA   NA   NA   NA                  NA -1.24920452 -1.097619534
[24,] NA   NA   NA   NA   NA   NA                  NA -1.17536151 -1.127874255
[25,] NA   NA   NA   NA   NA   NA                  NA -1.03222525 -0.958126485
[26,] NA   NA   NA   NA   NA   NA                  NA -0.74122697 -0.859642327
[27,] NA   NA   NA   NA   NA   NA                  NA -0.46319559 -0.705830514
[28,] NA   NA   NA   NA   NA   NA                  NA -0.34032953 -0.554033041
[29,] NA   NA   NA   NA   NA   NA                  NA -0.13787335 -0.366101772
[30,] NA   NA   NA   NA   NA   NA                  NA          NA -0.153727695
[31,] ⋮    ⋮    ⋮    ⋮    ⋮    ⋮           ⋮          ⋮           ⋮           
[32,] NA   NA   NA   NA   NA            NA         NA         NA  -1.03765130 
[33,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[34,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[35,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[36,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[37,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[38,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[39,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[40,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[41,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[42,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[43,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[44,] NA   NA   NA   NA   NA            NA         NA         NA           NA 
[45,] NA   NA   NA   NA   NA            NA -0.1697439 -0.4409237  -0.55458266 
[46,] NA   NA   NA   NA   NA   -0.06098633 -0.4206600 -0.7029645  -0.78887171 
[47,] NA   NA   NA   NA   NA   -0.15486561 -0.5678198 -0.4074512  -0.72860694 
[48,] NA   NA   NA   NA   NA   -0.24057946 -0.6678612 -0.9029381  -0.94882452 
[49,] NA   NA   NA   NA   NA   -0.22199807 -0.8262340 -1.3077309  -1.35774004 
[50,] NA   NA   NA   NA   NA   -0.36582714 -0.9412854 -1

But in most cases, we want to select based on time (or whichever coordinate), not based on the index. So let's now select based on a given time. The units for time are

In [17]:
att.get.nc(data, "time", "units")

[1] "days since 1800-01-01 00:00:00"

There are specific recommendations on how time should be stored in CF-NetCDF files. I will try to explain briefly here, and there is a nice explanation here too: https://www.unidata.ucar.edu/software/netcdf/time/recs.html

The *time* variable has units that count from a user defined origin, for example "hours since 2020-01-01 00:00 UTC" or "days since 2014-01-01". The units may be in years, days, seconds, nanoseconds, etc. Whilst this approach may seem strange at a glance, it allows the times to be stored in conventional numerical formats such as integers or floats, and to our desired precision. This is much more efficient than using a long timestamp string for each coordinate.

Some softwares know how to interpret this and will convert the data into timestamps in when you extract the data from a CF-NetCDF file. For example, xarray in Python does this, so does Panoply. Unfortunately, RNetCDF does not - at least not at the time of writing.

So now let's extract the time series, work out what index in the time series our desired date corresponds to, and then select from the *anom* variable based on that index.

In [18]:
desired_date <- as.Date('2020-01-01')
days_since_1800 <- as.numeric(difftime(desired_date, as.Date('1800-01-01'), units = 'days'))

time <- var.get.nc(data, "time") 
# Finding index of the value
time_index <- which(time == days_since_1800)

anom <- var.get.nc(data, 'anom', start=c(NA, NA, 1, time_index), count=c(NA,NA,1,1))
anom

       [,1]  [,2]  [,3]       [,4]         [,5]         [,6]          [,7]
 [1,]    NA    NA    NA -0.0900000 -0.477414876 -0.285908669 -0.1473901421
 [2,]    NA    NA    NA         NA -0.513699830 -0.280149400 -0.1605455577
 [3,]    NA    NA    NA  0.3000000 -0.426729143 -0.275595844 -0.1837482005
 [4,]    NA    NA    NA         NA -0.272711098 -0.233452842 -0.1684388518
 [5,]    NA    NA    NA         NA -0.180816099 -0.169976503 -0.1139785573
 [6,]    NA    NA    NA         NA -0.104851246 -0.083154038 -0.0783889070
 [7,]    NA    NA    NA         NA -0.203997135 -0.037347343 -0.0506733395
 [8,]    NA    NA  2.24         NA  0.409999996  0.049840331 -0.0306473132
 [9,]    NA    NA    NA  1.8500000           NA  0.096186161 -0.0211993735
[10,]    NA    NA    NA         NA  0.079999998  0.093276709  0.0013234944
[11,]    NA    NA    NA         NA           NA  0.101126052  0.0235306304
[12,]    NA    NA    NA         NA           NA  0.155259356  0.0395162962
[13,]    NA    NA    NA  

If you don't know the variable name, you can use the index instead.

In [19]:
var.inq.nc(data, 2)

$id
[1] 2

$name
[1] "lon"

$type
[1] "NC_FLOAT"

$ndims
[1] 1

$dimids
[1] 1

$natts
[1] 10

## Citing data

Remember, if you use a scientific dataset in a publication you should cite it in the same way you cite a paper - in the list of references. You can also include a statement in the data availability statement, but this should be as well as citing the dataset in the list of references.

This is the recommended citation for the data used in this tutorial - remember to update the data accessed:

H.-M. Zhang, B. Huang, J. H. Lawrimore, M. J. Menne, and T. M. Smith (2019): NOAA Global Surface Temperature Dataset (NOAAGlobalTemp), Version 5.0. NOAA National Centers for Environmental Information. doi:10.25921/9qth-2p70 Accessed 2024-04-10.